stat/distuv/pareto.go

```diff
@@ -0,0 +1,107 @@
+// Copyright ©2017 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distuv
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+)
+
+// Pareto implements the Pareto (Type I) distribution, a one parameter distribution
+// with support above the scale parameter.
+//
+// The density function is given by
+// \frac{\alpha x_m^{\alpha}}{x^{\alpha+1}} for x >= x_m
+//
+// For more information, see https://en.wikipedia.org/wiki/Pareto_distribution
+type Pareto struct {
+	// Xm is the scale parameter. Must be greater than 0.
+	Xm float64
+
+	// Alpha is the shape parameter. Must be greater than 0.
+	Alpha float64
+
+	Src *rand.Rand
+}
+
+// CDF computes the value of the cumulative density function at x.
+func (p Pareto) CDF(x float64) float64 {
+	if x < p.Xm {
+		return 0
+	}
+	return 1 - math.Pow(p.Xm/x, p.Alpha)
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (p Pareto) ExKurtosis() float64 {
+	if p.Alpha <= 4 {
+		return 0
+	}
+	return 6 * (math.Pow(p.Alpha, 3) + math.Pow(p.Alpha, 2) - 6*p.Alpha - 2) / p.Alpha / (p.Alpha - 3) / (p.Alpha - 4)
+
+}
+
+// LogProb computes the natural logarithm of the value of the probability
+// density function at x.
+func (p Pareto) LogProb(x float64) float64 {
+	if x < p.Xm {
+		return math.Inf(-1)
+	}
+	return math.Log(p.Alpha) + p.Alpha*math.Log(p.Xm) - (p.Alpha+1)*math.Log(x)
+}
+
+// Mean returns the mean of the probability distribution.
+func (p Pareto) Mean() float64 {
+	if p.Alpha > 1 {
+		return p.Alpha * p.Xm / (p.Alpha - 1)
+	} else {
+		return math.Inf(1)
+	}
+}
+
+// Mode returns the mode of the distribution.
+func (p Pareto) Mode() float64 {
+	return p.Xm
+}
+
+// NumParameters returns the number of parameters in the distribution.
+func (p Pareto) NumParameters() int {
+	return 2
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Pareto) Prob(x float64) float64 {
+	if x < p.Xm {
```

Comment:
Usually we `return math.Exp(p.LogProb(x))`.

Reply:
done

Reply:
Omit the condition, you already check for `x < p.Xm` in `LogProb` and return `-Inf` in that case; `math.Exp(math.Inf(-1))` returns 0.

Reply:
Ah, so obvious. Thank you. Done.